server/service/src/invoice/inventory_adjustment/adjust_existing_stock/insert.rs

```diff
@@ -54,19 +58,41 @@ pub fn insert_inventory_adjustment(
         .connection
         .transaction_sync(|connection| {
             let stock_line = validate(connection, &ctx.store_id, &input)?;
-            let (new_invoice, invoice_line, stock_line_row) =
-                generate(connection, &ctx.store_id, &ctx.user_id, input, stock_line)?;
+            let GenerateResult {
+                invoice,
+                insert_stock_in_or_out_line,
+                update_inventory_adjustment_reason,
+            } = generate(connection, &ctx.store_id, &ctx.user_id, input, stock_line)?;
 
+            // Create Inventory Adjustment in New status
             let invoice_row_repo = InvoiceRowRepository::new(connection);
+            invoice_row_repo.upsert_one(&invoice)?;
+
+            // Add invoice line (and update stock line)
+            match insert_stock_in_or_out_line {
+                InsertStockInOrOutLine::StockIn(stock_in_line) => {
+                    insert_stock_in_line(ctx, stock_in_line).map_err(|error| {
+                        InsertInventoryAdjustmentError::StockInLineInsertError(error)
+                    })?;
+                }
+                InsertStockInOrOutLine::StockOut(stock_out_line) => {
+                    insert_stock_out_line(ctx, stock_out_line).map_err(|error| {
+                        InsertInventoryAdjustmentError::StockOutLineInsertError(error)
+                    })?;
+                }
+            }
 
-            invoice_row_repo.upsert_one(&new_invoice)?;
-            InvoiceLineRowRepository::new(connection).upsert_one(&invoice_line)?;
-            StockLineRowRepository::new(connection).upsert_one(&stock_line_row)?;
+            // Add inventory adjustment reason to the invoice line
+            let invoice_line_repo = InvoiceLineRowRepository::new(&connection);
+            invoice_line_repo.update_inventory_adjustment_reason_id(
+                &update_inventory_adjustment_reason.invoice_line_id,
+                update_inventory_adjustment_reason.reason_id,
+            )?;
```

Comment:
I'm really not sure I'm convinced about this pattern for adding the reasons... part of me likes it - stock out/stock in should be shared logic, so things that are specific to one kind of invoice line (like a return reason, or an inventory adjustment reason) should be added outside of it.
But I feel like we're more likely to just end up with devs forgetting this step at some point?

Reply:
Hmm, I see. I kind of like the way it is now. I';ve been thinking about these common operations being abstracted as 'steps' and then composing services with 'steps' that are applicable, although not 100% sure what the balance is yet